bbq/compiler/codegen.go

```diff
@@ -0,0 +1,242 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Flow Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package compiler
+
+import (
+	"github.com/onflow/cadence/bbq/opcode"
+	"github.com/onflow/cadence/common"
+)
+
+type CodeGen interface {
+	Offset() int
+	Code() interface{}
+	EmitNil()
+	EmitTrue()
+	EmitFalse()
+	EmitDup()
+	EmitDrop()
+	EmitGetConstant(index uint16)
+	EmitJump(target uint16) int
+	EmitJumpIfFalse(target uint16) int
+	PatchJump(offset int, newTarget uint16)
+	EmitReturnValue()
+	EmitReturn()
+	EmitGetLocal(index uint16)
+	EmitSetLocal(index uint16)
+	EmitGetGlobal(index uint16)
+	EmitSetGlobal(index uint16)
+	EmitGetField()
+	EmitSetField()
+	EmitGetIndex()
+	EmitSetIndex()
+	EmitNewArray(index uint16, size uint16, isResource bool)
+	EmitIntAdd()
+	EmitIntSubtract()
+	EmitIntMultiply()
+	EmitIntDivide()
+	EmitIntMod()
+	EmitEqual()
+	EmitNotEqual()
+	EmitIntLess()
+	EmitIntLessOrEqual()
+	EmitIntGreater()
+	EmitIntGreaterOrEqual()
+	EmitUnwrap()
+	EmitCast(index uint16, kind opcode.CastKind)
```

Comment:
Same here, and some more instructions below, like `EmitTransfer`, `EmitNewRef`, `EmitNew`
```suggestion
	EmitCast(typeIndex uint16, kind opcode.CastKind)
```

Reply:
Improved here and in the other cases in [d02289b](https://github.com/onflow/cadence/pull/3713/commits/d02289b09a0f92570eabead6c3751b3b1cfa05ef)